Advanced settings: Smaller int/float formspec. Use 'textarea'

paramat opened the pull request:
Advanced settings: Smaller int/float formspec. Use 'textarea'

Previously, the int/float editing formspec displayed larger than the
screen on many Android devices.
Reduce its size to make it possible for Android users to edit the
critical GUI/HUD scaling and font size settings.
For all value-editing formspecs use 'textarea' to enable line-wrapping,
to avoid clipped comment lines.
//////////////////////////

![screenshot from 2018-10-19 06-14-30](https://user-images.githubusercontent.com/3686677/47199174-2e33b400-d368-11e8-85f3-c5d24a93dfcb.png)

^ Int/float edit size reduced from 10 x 5.2 to 8 x 4.23. 'blah' lines added to show scrollbar and scrolling ability.

![screenshot from 2018-10-17 08-07-33](https://user-images.githubusercontent.com/3686677/47199674-a602de00-d36a-11e8-84ba-3b14aa57b7d9.png)

^ How it looks in MT master (much worse, too much space)

![screenshot from 2018-10-19 06-20-55](https://user-images.githubusercontent.com/3686677/47199176-312ea480-d368-11e8-8d26-35dd6a58fbae.png)

^ Noise edit shown because the '(2D Noise)' label needed moving downwards a little.

See #7784 for more details and discussion.
Also closes #3808 

https://github.com/minetest/minetest/pull/7784#issuecomment-429655579
"I think that the important thing is that the UI is always at least useable enough to alter the settings to your own personal preference."

https://github.com/minetest/minetest/pull/7784#issuecomment-430494630
"the main problem here is the oversized advanced settings editing formspec.
I might alter the code such that the formspec for editing simple settings (GUI/HUD scaling, font size) is as small as possible, then this can be a short term solution to at least make it possible to edit these values on any Android device.
There are formspecs for editing noise params, mapgen flags etc. which need to be large so i'll leave those unchanged."

The Android app is usable once the user sets GUI/HUD scaling and font size. However adjusting those values is difficult due to the advanced settings edit formspec often being much larger than the screen.
This PR is intended to make those settings easier to set.

For desktop this also helps by not having an excessively sized formspec for setting ints and floats.

SmallJoker commented on builtin/mainmenu/dlg_settings_advanced.lua:
```diff
@@ -532,20 +532,31 @@ local checkboxes = {} -- handle checkboxes events
 			"tablecolumns[color;text]" ..
 			"tableoptions[background=#00000000;highlight=#00000000;border=false]" ..
-			"table[0,0;10,3;info;"
+			"table[0,0;" .. width .. "," .. iheight .. ";info;"
```
Could you please make use of textareas here? Otherwise users will again complain when their font is too large to display the description lines.

paramat replied:
I have to learn how, working on it.

paramat replied:
Done.

stujones11 commented:
I have a feeling this might be a bit too drastic a reduction for some, however, it would surely help with the android problem. Maybe just a little bit bigger? Or just use a textarea as @SmallJoker  suggested.

paramat commented:
Updated to use textarea, added a dark box too. See new screenshots in first post.
Size could be tuned, easy to do by altering the constants.

paramat commented:
Hang on, i've wasted some space ...

paramat commented:
Ok updated again. Wider at 7.5 and height reduced to not waste space, updated first screenshot.

paramat commented:
Hm i need to not have the setting name so close to the top of the box, WIP.

paramat commented:
Updated again, new screenshots.

stujones11 commented:
Looks good overall :+1: 

Although, in your second screenshot, do those edit-boxes really need to be that wide?

SmallJoker commented:
:-1: Yet not keen on this cramped and tiny formspec on the desktop. The width/height calculation looks very hacky and should be replaced by adding the `size[` element shortly before return, same as the buttons which then would be moved down, depending on the setting type.

I will come up with a code suggestion as soon (hopefully tomorrow or on the weekend).

stujones11 commented:
Code changes aside, I actually prefer this on desktop, I think the textarea is much clearer with the dark background. I guess it depends on whether you play in a small windowed mode or in near/full screen.

paramat commented:
I'm not happy with the setting title being the label of the textarea, i'll make it a separate element and make it yellow as before, and make it stand out more.

paramat commented:
Updated, see new screenshots.

The setting name was cramped, i moved it into a separate label with plenty of space around it.
Int/float edit formspec is now width 8 (previously 10) so is not particularly small.
There was some unnecessary code, removed 'iheight' variable.
The width variable makes it easy infuture to tune the width according to the requirements of Android apps. Hopefully width 8 is not too big for Android, we'll see.

SmallJoker commented:
The diff is quite large (11 KiB), based on clean master: https://krock-works.uk.to/u/new_adv_settings_2_fix_error_box.diff

**What's changed:**
1) Formspec size and description box are calculated last
2) Width and height are now adjustable per setting type
3) Error message (`dialogdata.error_message`) shortens the description field and is placed below
4) Add more spacing for larger fonts
5) More comments and extensible by setting different height and width values

paramat commented:
Ok this is likely much better than my attempt, could you create a branch with this and open a PR for easier review and testing?

paramat commented:
@SmallJoker 2D noise uses spread components X and Z.

paramat commented:
Replaced by #7802
